src/internal/modules/invoice-accounts/controllers/invoice-accounts.js

```diff
@@ -22,27 +23,28 @@ const helpers = require('../lib/helpers');
 const getCompany = async (request, h) => {
   const { regionId, companyId } = request.params;
   const { licenceId, redirectPath } = request.query;
-  const { company } = request.pre;
+  const { company, companies } = request.pre;
   // licenceId is an optional query param. if supplied it will be displayed as the caption on the forms for the session
   const { licenceNumber } = licenceId ? await dataService.getLicenceById(licenceId) : { licenceNumber: null };
   // The company name and licence number set here will be used in the select address page
   const data = { viewData: { redirectPath, licenceNumber, licenceId, companyName: titleCase(company.name) } };
   const session = dataService.sessionManager(request, regionId, companyId, data);
-  const selectedCompany = has(session, 'agent') && isEmpty(session.agent) ? company : helpers.getAgentCompany(session);
+  const selectedCompany = has(session, 'agent') && isEmpty(session.agent) ? company : await helpers.getAgentCompany(session);
+
   return h.view('nunjucks/form', {
     ...request.view,
     caption: helpers.getFormTitleCaption(licenceNumber),
     pageTitle: 'Who should the bills go to?',
     back: '/manage',
-    form: sessionForms.get(request, selectCompanyForm(request, company, selectedCompany))
+    form: sessionForms.get(request, selectCompanyForm(request, companies, selectedCompany))
   });
 };
 
 const postCompany = async (request, h) => {
   const { regionId, companyId } = request.params;
-  const { company } = request.pre;
+  const { companies } = request.pre;
```

Comment:
if you decide to get rid of the loadCompanies pre-handler then you can do the same here and only add the agent company if it has been selected. i.e. if the payload.selected company id != company Id then get it using the helpers.getAgentCompany.

Reply:
I have slimmed this down by removing one of the pre-handlers.